File: app/src/main/java/com/revosleap/bxplayer/Fragments/FragmentTracks.java
package com.revosleap.bxplayer.Fragments;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.revosleap.bxplayer.AppUtils.BxPlayback.BXNotificationManager;
import com.revosleap.bxplayer.AppUtils.BxPlayback.BxPlayerService;
import com.revosleap.bxplayer.AppUtils.BxPlayback.PlaybackInfoListener;
import com.revosleap.bxplayer.AppUtils.BxPlayback.PlayerAdapter;
import com.revosleap.bxplayer.AppUtils.Models.AudioModel;
import com.revosleap.bxplayer.AppUtils.RecyclerView.Adapters.TrackAdapter;
import com.revosleap.bxplayer.AppUtils.RecyclerView.Listeners.CustomTouchListener;
import com.revosleap.bxplayer.AppUtils.Player.AudioPlayerService;
import com.revosleap.bxplayer.AppUtils.Utils.GetAudio;
import com.revosleap.bxplayer.AppUtils.Utils.StorageUtil;
import com.revosleap.bxplayer.AppUtils.Utils.onItemClickListener;
import com.revosleap.bxplayer.R;

import java.util.ArrayList;
import java.util.List;


public class FragmentTracks extends Fragment implements TrackAdapter.SongSelectedListener {

    RecyclerView recyclerView;
    private AudioPlayerService player;
    boolean serviceBound = false;
    List<AudioModel>list;
    public static final String Broadcast_PLAY_NEW_AUDIO = "com.revosleap.bxplayer.PlayNewAudio";
    BxPlayerService mMusicService;
    private boolean mIsBound;
    private TrackAdapter adapter;
    private PlayerAdapter mPlayerAdapter;
    private BXNotificationManager mMusicNotificationManager;
    private PlaybackListener mPlaybackListener;
    private ServiceConnection mConnection= new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            mMusicService=((BxPlayerService.BXBinder)service).getInstance();
            mPlayerAdapter = mMusicService.getMediaPlayerHolder();
            mMusicNotificationManager = mMusicService.getMusicNotificationManager();
            if (mPlaybackListener == null) {
                mPlaybackListener = new PlaybackListener();
                mPlayerAdapter.setPlaybackInfoListener(mPlaybackListener);
            }
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            mMusicService = null;
        }
    };





    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {


        View view= inflater.inflate(R.layout.fragment_tracks, container, false);


        list= new GetAudio().geAllAudio(getActivity());
        adapter= new TrackAdapter(list,getActivity(),this);

        recyclerView = view.findViewById(R.id.trackRecycler);
        //TrackAdapter adapter= new TrackAdapter(list,getActivity());
        recyclerView.setAdapter(adapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
//        recyclerView.addOnItemTouchListener(new CustomTouchListener(getActivity(), new onItemClickListener() {
//            @Override
//            public void onClick(View view, int index) {
//               playAudio(index);
//            }
//        }));

        return view;


    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        doBindService();
    }

    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            // We've bound to LocalService, cast the IBinder and get LocalService instance
            AudioPlayerService.LocalBinder binder = (AudioPlayerService.LocalBinder) service;
            player = binder.getService();
            serviceBound = true;
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            serviceBound = false;
        }
    };

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean("serviceStatus",serviceBound);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
//        serviceBound=savedInstanceState.getBoolean("serviceStatus");
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        super.onDestroy();
        mPlaybackListener = null;
        doUnbindService();
    }

    private void playAudio(int audioIndex) {
        ArrayList<AudioModel>arrayList = null;
        //Check is service is active
        if (!serviceBound) {
            //Store Serializable audioList to SharedPreferences
            StorageUtil storage = new StorageUtil(getActivity());
            storage.storeAudio(list);
            storage.storeAudioIndex(audioIndex);

            Intent playerIntent = new Intent(getActivity(), AudioPlayerService.class);
            getActivity().startService(playerIntent);
            getActivity().bindService(playerIntent, serviceConnection, Context.BIND_AUTO_CREATE);
        } else {
            //Store the new audioIndex to SharedPreferences
            StorageUtil storage = new StorageUtil(getActivity());
            storage.storeAudioIndex(audioIndex);

            //Service is active
            //Send a broadcast to the service -> PLAY_NEW_AUDIO
            Intent broadcastIntent = new Intent(Broadcast_PLAY_NEW_AUDIO);
            getActivity().sendBroadcast(broadcastIntent);
        }
    }


    @Override
    public void onSongSelected(@NonNull AudioModel song, @NonNull List<AudioModel> songs) {
        Log.v("song ",song.getTitle()+" number= "+songs.size());
        try {
            mPlayerAdapter.setCurrentSong(song, songs);
            mPlayerAdapter.initMediaPlayer();
        }catch (Exception e){
            e.printStackTrace();
        }
        mPlayerAdapter.getMediaPlayer().start();
        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                mMusicService.startForeground(BXNotificationManager.NOTIFICATION_ID,
                        mMusicNotificationManager.createNotification());
            }
        }, 250);

    }
    private void doBindService() {
        // Establish a connection with the service.  We use an explicit
        // class name because we want a specific service implementation that
        // we know will be running in our own process (and thus won't be
        // supporting component replacement by other applications).
        getActivity().bindService(new Intent(getActivity(),
                BxPlayerService.class), mConnection, Context.BIND_AUTO_CREATE);
        mIsBound = true;

        final Intent startNotStickyIntent = new Intent(getActivity(),BxPlayerService.class);
        getActivity().startService(startNotStickyIntent);
    }
    private void doUnbindService() {
        if (mIsBound) {
            // Detach our existing connection.
            getActivity().unbindService(mConnection);
            mIsBound = false;
        }
    }
    class PlaybackListener extends PlaybackInfoListener {

        @Override
        public void onPositionChanged(int position) {
//            if (!mUserIsSeeking) {
//                mSeekBarAudio.setProgress(position);
//            }
        }

        @Override
        public void onStateChanged(@State int state) {
//
//            updatePlayingStatus();
//            if (mPlayerAdapter.getState() != State.RESUMED && mPlayerAdapter.getState() != State.PAUSED) {
//                updatePlayingInfo(false, true);
//            }
        }

        @Override
        public void onPlaybackCompleted() {
            //updateResetStatus(true);
        }
    }
}

File: app/src/main/java/com/revosleap/bxplayer/ui/fragments/FragmentTracks.kt
package com.revosleap.bxplayer.ui.fragments

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.revosleap.bxplayer.R
import com.revosleap.bxplayer.services.MusicPlayerService
import com.revosleap.bxplayer.utils.adapters.TrackAdapter
import com.revosleap.bxplayer.models.AudioModel
import com.revosleap.bxplayer.utils.playback.BXNotificationManager
import com.revosleap.bxplayer.utils.playback.PlaybackInfoListener
import com.revosleap.bxplayer.utils.playback.PlayerAdapter
import com.revosleap.bxplayer.utils.utils.GetAudio
import kotlinx.android.synthetic.main.fragment_tracks.*


class FragmentTracks : Fragment(), TrackAdapter.SongSelectedListener {


    private var serviceBound = false
    private var list = mutableListOf<AudioModel>()
    private var mMusicService: MusicPlayerService? = null
    private var mIsBound: Boolean = false
    private var trackAdapter: TrackAdapter? = null
    private var mPlayerAdapter: PlayerAdapter? = null
    private var mMusicNotificationManager: BXNotificationManager? = null
    private var mPlaybackListener: PlaybackListener? = null
    private val mConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            mMusicService = (service as MusicPlayerService.MusicBinder).serviceInstance
            mPlayerAdapter = mMusicService!!.mediaPlayerHolder
            mMusicNotificationManager = mMusicService!!.musicNotificationManager
            if (mPlaybackListener == null) {
                mPlaybackListener = PlaybackListener()
                mPlayerAdapter!!.setPlaybackInfoListener(mPlaybackListener!!)
            }
        }

        override fun onServiceDisconnected(name: ComponentName) {
            mMusicService = null
        }
    }


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {


        val view = inflater.inflate(R.layout.fragment_tracks, container, false)


        list = GetAudio().geAllAudio(activity!!.baseContext)
        trackAdapter = TrackAdapter(list, activity!!, this)

        return view


    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        trackRecycler.apply {
            adapter = trackAdapter
            layoutManager = LinearLayoutManager(activity)
            hasFixedSize()
        }
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        doBindService()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putBoolean("serviceStatus", serviceBound)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        super.onViewStateRestored(savedInstanceState)
        //        serviceBound=savedInstanceState.getBoolean("serviceStatus");
    }

    override fun onDestroy() {
        super.onDestroy()
        super.onDestroy()
        mPlaybackListener = null
        doUnbindService()
    }

    override fun onSongSelected(song: AudioModel, songs: List<AudioModel>) {
        mPlayerAdapter!!.setCurrentSong(song, songs)
        mPlayerAdapter!!.initMediaPlayer()
        mPlayerAdapter!!.getMediaPlayer()?.start()
        mMusicService!!.startForeground(BXNotificationManager.NOTIFICATION_ID,
                mMusicNotificationManager!!.createNotification())


    }

    private fun doBindService() {
        activity!!.bindService(Intent(activity,
                MusicPlayerService::class.java), mConnection, Context.BIND_AUTO_CREATE)
        mIsBound = true

        val startNotStickyIntent = Intent(activity, MusicPlayerService::class.java)
        activity!!.startService(startNotStickyIntent)
    }

    private fun doUnbindService() {
        if (mIsBound) {
            // Detach our existing connection.
            activity!!.unbindService(mConnection)
            mIsBound = false
        }
    }

    internal inner class PlaybackListener : PlaybackInfoListener() {

        override fun onPositionChanged(position: Int) {
            //            if (!mUserIsSeeking) {
            //                mSeekBarAudio.setProgress(position);
            //            }
        }

        override fun onStateChanged(@State state: Int) {
            //
            //            updatePlayingStatus();
            if (mPlayerAdapter?.getState() != State.RESUMED && mPlayerAdapter?.getState() != State.PAUSED) {
                updatePlayingInfo(false, true);
            }
        }

        override fun onPlaybackCompleted() {
            //updateResetStatus(true);
        }
    }

    private fun updatePlayingInfo(restore: Boolean, startPlay: Boolean) {
        if (startPlay) {
            mPlayerAdapter!!.getMediaPlayer()?.start()
            Handler().postDelayed({
                mMusicService!!.startForeground(BXNotificationManager.NOTIFICATION_ID,
                        mMusicNotificationManager!!.createNotification())
            }, 250)
        }
        if (restore) {
            Handler().postDelayed({
                //stop foreground if coming from pause state
                if (mMusicService!!.isRestoredFromPause) {
                    mMusicService!!.stopForeground(false)
                    mMusicService!!.musicNotificationManager?.notificationManager!!
                            .notify(BXNotificationManager.NOTIFICATION_ID,
                                    mMusicService!!.musicNotificationManager?.notificationBuilder?.build())
                    mMusicService!!.isRestoredFromPause = false
                }
            }, 250)
        }
    }

    companion object {
        const val Broadcast_PLAY_NEW_AUDIO = "com.revosleap.bxplayer.PlayNewAudio"
    }
}
